types/wlr_relative_pointer_v1.c

```diff
@@ -0,0 +1,261 @@
+#include <stdlib.h>
+#include <assert.h>
+#include <inttypes.h>
+#include <wlr/util/log.h>
+#include <util/signal.h>
+#include <wlr/types/wlr_relative_pointer_v1.h>
+#include <wlr/types/wlr_seat.h>
+#include "wayland-util.h"
+#include "wayland-server.h"
+#include "relative-pointer-unstable-v1-protocol.h"
+
+#define RELATIVE_POINTER_MANAGER_VERSION 1
+
+static const struct zwp_relative_pointer_manager_v1_interface relative_pointer_manager_v1_impl;
+static const struct zwp_relative_pointer_v1_interface relative_pointer_v1_impl;
+
+
+/**
+ * helper functions
+ */
+
+struct wlr_relative_pointer_v1 *wlr_relative_pointer_v1_from_resource(struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zwp_relative_pointer_v1_interface,
+		&relative_pointer_v1_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+
+static struct wlr_relative_pointer_manager_v1 *relative_pointer_manager_from_resource(struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zwp_relative_pointer_manager_v1_interface,
+		&relative_pointer_manager_v1_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+
+/**
+ * relative_pointer handler functions
+ */
+
+static void relative_pointer_destroy(struct wlr_relative_pointer_v1 *relative_pointer) {
+	wlr_signal_emit_safe(&relative_pointer->events.destroy, relative_pointer);
+
+	wl_list_remove(&relative_pointer->link);
+	wl_list_remove(&relative_pointer->seat_destroy.link);
+	free(relative_pointer);
+}
+
+static void relative_pointer_v1_handle_resource_destroy(struct wl_resource *resource) {
+	struct wlr_relative_pointer_v1 *relative_pointer =
+		wlr_relative_pointer_v1_from_resource(resource);
+	if (relative_pointer == NULL) {
+		return;
+	}
+	relative_pointer_destroy(relative_pointer);
+}
+
+
+static void relative_pointer_v1_handle_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	struct wlr_relative_pointer_v1 *relative_pointer =
+		wlr_relative_pointer_v1_from_resource(resource);
+	wlr_log(WLR_DEBUG, "relative_pointer_v1 %p released by client %p",
+		relative_pointer, client);
+
+	wl_resource_destroy(resource);
+}
+
+static void relative_pointer_handle_seat_destroy(struct wl_listener *listener,
+		void *data) {
+	struct wlr_relative_pointer_v1 *relative_pointer =
+		wl_container_of(listener, relative_pointer, seat_destroy);
+
+	relative_pointer_destroy(relative_pointer);
+	wl_resource_set_user_data(relative_pointer->resource, NULL);
```

Comment:
This is a use-after-free.

Reply:
Actually, this instruction (which makes the resource inert) should be moved in `relative_pointer_destroy`.

Reply:
Oh shoot you're right.  I can't believe I didn't see that.  Will do